In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error

In [2]:
import seaborn as sns
df = sns.load_dataset("titanic")
df = df.dropna(subset=["fare", "age", "pclass"])
df.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [3]:
X = df[["age", "pclass"]]
y = df["fare"]

print(X.shape, y.shape)

(714, 2) (714,)


In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(X_train.shape, X_test.shape)

(571, 2) (143, 2)


In [5]:
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

lr_predictions = lr_model.predict(X_test)
lr_mae = mean_absolute_error(y_test, lr_predictions)

print("Linear Regression MAE:", lr_mae)

Linear Regression MAE: 24.322781125639057


In [6]:
rf_model = RandomForestRegressor(random_state=42)
rf_model.fit(X_train, y_train)

rf_predictions = rf_model.predict(X_test)
rf_mae = mean_absolute_error(y_test, rf_predictions)

print("Random Forest MAE:", rf_mae)

Random Forest MAE: 23.86184549422067


In [7]:
print("Linear Regression MAE:", lr_mae)
print("Random Forest MAE:", rf_mae)
print("Difference:", lr_mae - rf_mae)

Linear Regression MAE: 24.322781125639057
Random Forest MAE: 23.86184549422067
Difference: 0.460935631418387


## Model Comparison: Linear Regression vs Random Forest

**Task:** Predict `fare` from `age` and `pclass` (Titanic dataset)

**Results:**
- Linear Regression MAE: $24.32
- Random Forest MAE: $23.86

**Conclusion:**
Random Forest performed slightly better than Linear Regression on this data 
(MAE of $23.86 vs $24.32), though the difference was modest. This makes sense 
because fare doesn't have a purely straight-line relationship with age and 
class — fare varies a lot even within the same class, and Random Forest can 
capture those more complex, non-linear patterns better than a single straight 
line can.

That said, since the improvement was small, Linear Regression could still be 
the better choice in a setting where interpretability matters more than a 
small accuracy gain — its coefficients are easy to explain directly, while 
Random Forest's many combined trees are harder to interpret.